lib/Elevate/Components/CCS.pm

```diff
@@ -0,0 +1,424 @@
+
+    my $delegate_file = $path . '/' . 'delegates.json';
+    Cpanel::JSON::DumpFile( $delegate_file, $delegates_ar );
+    return;
+}
+
+sub _dump_persistence_data_for_user ( $self, $user ) {
+    my $path             = $self->_get_export_path_for_user($user);
+    my $persistence_file = $path . '/' . 'persistence.json';
+    my $users_ccs_info   = $self->_get_ccs_info_for_user($user);
+
+    Cpanel::JSON::DumpFile( $persistence_file, $users_ccs_info );
+    return;
+}
+
+sub _make_backup_paths_for_user ( $self, $user ) {
+    my $path = $self->_get_export_path_for_user($user);
+    File::Path::make_path($path);
+
+    my %dump_types = DUMP_TYPES();
+    for ( keys(%dump_types) ) { File::Path::make_path("$path/$_"); }
+    return;
+}
+
+sub _get_query_for_type ( $self, $type, $uuid ) {
+    my %querydata = (
+        'calendars' => {
+            'args'  => [ $uuid, '1', 'f' ],
+            'query' => "SELECT icalendar_text
+                FROM
+                  calendar_object
+                  INNER JOIN calendar_bind ON calendar_bind.calendar_resource_id = calendar_object.calendar_resource_id
+                  INNER JOIN calendar_metadata ON calendar_metadata.resource_id = calendar_bind.calendar_resource_id
+                  INNER JOIN calendar_home ON calendar_home.resource_id = calendar_bind.calendar_home_resource_id
+                WHERE calendar_home.owner_uid = ?
+                AND calendar_bind.bind_status = ?
+                AND calendar_metadata.is_in_trash = ?;",
+        },
+        'contacts' => {
+            'args'  => [ $uuid, 'f' ],
+            'query' => "SELECT vcard_text
+                FROM
+                  addressbook_object
+                  INNER JOIN addressbook_home ON addressbook_home.resource_id = addressbook_object.addressbook_home_resource_id
+                WHERE addressbook_home.owner_uid = ?
+                AND addressbook_object.is_in_trash = ?;",
+        },
+    );
+
+    return ( $querydata{$type}{'query'}, $querydata{$type}{'args'} );
+}
+
+sub _get_dbh ($self) {
+    $self->{dbh} ||= Cpanel::CCS::DBUtils::get_dbh();
+    return $self->{dbh};
+}
+
+sub _get_export_path_for_user ( $self, $user ) {
+    $self->{$user}{export_path} ||= EXPORT_DIR() . '/' . $user . '/calendar_and_contacts';
+    return $self->{$user}{export_path};
+}
+
+sub _get_ccs_info_for_user ( $self, $user ) {
+    $self->{$user}{info} ||= Cpanel::CCS::Userdata::get_cpanel_account_users_uuids($user);
+    return $self->{$user}{info};
+}
+
+sub _ensure_export_directory ($self) {
+    File::Path::make_path(EXPORT_DIR);
+    return;
+}
+
+####################################
+##### post_leapp code below this ###
+####################################
+
+sub post_leapp ($self) {
+    return unless Elevate::StageFile::read_stage_file('ccs_installed');
+
+    $self->dnf->install(CCS_PACKAGE);
+    $self->_ensure_ccs_service_is_up();
+    $self->_clear_task_queue();
+    $self->run_once('import_ccs_data');
```

Comment:
What if the CCS service isn't up. don't we need to fail and abort the script?

Reply:
Oh, I see, we're trying to gracefully fail and move on?

Reply:
Correct, it should fail gracefully, and give a final notification of any users that failed at the end of the log.  It will also display the backup directory and recommend contacting support if they need further assistance importing the data at that point.

I chose to do this instead of failing as I don't think that failing to import the CCS should be a fatal error for the script since the data can still be recovered.  I think this is YAGNI for now, but we could expose an option for elevated servers to import the data if it becomes enough of a problem.